### PR TITLE
NEO-2115 // update Table row selection

### PR DESCRIPTION
**Pull request by joe-s-avaya**
[Link to updated Table story](https://deploy-preview-444--neo-react-library-storybook.netlify.app/?path=/story/components-table--editable-data)

**Before tagging the team for a review, I have done the following:**

- [x] run `yarn all` locally: ensures that all tests pass, formatting is done, types pass, and builds pass
- [x] reviewed my code changes
- [x] updated the link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [x] done an accessibility check on my work (tested with Chrome's `axe Dev Tools`, Mac's VoiceOver, etc.)
- [x] tagged `@avaya-dux/dux-design` if any visual changes have occurred
- [x] tagged `@avaya-dux/dux-devs`

This makes several updates to row selection in the Table component. Most notably that the header checkbox now selects/deselects the _page's_ rows, _not_ the entire tables.

[Figma link](https://www.figma.com/file/o62kJbRQpFq9T0fGwKvCMA/Component-Exploration-FY22---v3?type=design&node-id=3313-531108&mode=design&t=1IRQvmML9xvf4Oo0-0)

`@avaya-dux/dux-design`: Review table selection visuals.

- "Select all" checkbox appropriately appears and updates.
- Chevron next to the checkbox is visually appropriate (sizing, spacing, etc.).

`@avaya-dux/dux-devs`: Review table selection logic.

- review code and test logic
- review [use-cases spreadsheet](https://avaya365-my.sharepoint.com/:x:/g/personal/jsebast_avaya_com/EaYuTO3lFF9DiKXE1jcrCPcBR7VnZojp50o4zXPv_JNy1Q?e=Q8jXVg) and ensure all use-cases are handled
- review updates to Menu components and stories to ensure nothing has inappropriately changed. 
  - IMPORTANT: all current "Menu Item" components will not be fully a11y due to them not being `<button>`s. We'll need to figure out how to fix that soon.

## Summary by CodeRabbit

- **New Features**
  - Enhanced row selection functionality in tables, including selecting visible rows, clearing selections, and toggling row states.
- **Improvements**
  - Updated table header translations for better clarity on selection actions.
- **Bug Fixes**
  - Simplified row selection logic in test cases to improve reliability.
  - Improved checkbox label references for better test accuracy.

### CODE REVIEW

**Comment by yangAtSpoken** [comment]
Screenshot 2024-07-09 at 1 14 10 PM

Use Case review:  the last use case does not seem working. My understand of this use case is: when all rows cross all pages are selected, clicking a header checkbox of any page, will de-select all items cross all pages.  What actually happens is  only rows of the clicked page are deselected.

**Comment by yangAtSpoken** [comment]
functional

Can not move focus to the last menu item with keyboard.

File: src/components/Table/Table.test.jsx [hunk 1]
```diff
@@ -315,16 +312,21 @@ describe("Table", () => {
...
 
-			const deleteButton = screen.getByText("Delete");
-			await user.click(deleteButton);
+			// delete first page
+			await user.click(screen.getByLabelText(selectAllCheckboxLabel));
+			await user.click(screen.getByText("Delete"));
+
+			// delete second (final) page
+			await user.click(screen.getByLabelText(selectAllCheckboxLabel));
+			await user.click(screen.getByText("Delete"));
```
Comment (yangAtSpoken) [inline src/components/Table/Table.test.jsx:325]: nice

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 2]
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
...
+		page,
 		rows,
 		rowsById,
 		state: { selectedRowIds },
+		toggleRowSelected,
 		toggleAllRowsSelected,
 		toggleSortBy,
 	} = instance;
 
+	const [visibleRows, allVisibleRowsAreSelected, allVisibleRowsDeselected] =
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:44]: "visible" and "shown" are not accurate to me. If the page is not selected, all rows in that page are not visible/shown to the user.

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 3]
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
...
+		useMemo(() => {
+			const shownRows = page.map((row) => row.original);
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:46]: shownRows => pageRows

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 4]
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
...
+			const visibleRowsSelected =
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:47]: allRowsSelected or allPageRowsSelected

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 5]
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
...
+				shownRows.length && shownRows.every((row) => selectedRowIds[row.id]);
+			const visibleRowsNotSelected = shownRows.every(
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:49]: allRowsNoteSlected or noRowsSelected

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 6]
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
...
+				(row) => !selectedRowIds[row.id],
+			);
+
+			return [shownRows, visibleRowsSelected, visibleRowsNotSelected];
+		}, [page, selectedRowIds]);
+
+	const selectVisibleRows = useCallback(
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:56]: selectPageRows

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 7]
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
...
+		(set: boolean) => {
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:57]: selected: bolean

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 8]
```diff
@@ -69,19 +91,73 @@ export const TableHeader = <T extends Record<string, any>>({
 		checkboxCheckedValue,
 		rowsById,
 	]);
+	const toggleAllVisibleRows = useCallback(() => {
```
Comment (yangAtSpoken) [inline src/components/Table/TableComponents/TableHeader.tsx:94]: togglePageRows

**Review by yangAtSpoken — APPROVED** [review]
The Table update looks great to me despite of all the naming suggestions.  However, I do not see that we need a MenuItemButton component as MenuItem has onClick callback already.

File: src/components/Menu/MenuItemButton/MenuItemButton.tsx [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import clsx from "clsx";
+
+export interface MenuItemButtonProps
+	extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+	children?: React.ReactNode;
+}
+
+export const MenuItemButton = ({
```
[the remark this answers is not in the document]
Comment (joe-s-avaya) [reply src/components/Menu/MenuItemButton/MenuItemButton.tsx:8]: The original `MenuItem` is a `div`, which does not have the built-in a11y that the `button` has. However, due to our `Menu` assuming that we're handling all the things, I'm running into quite a few issues with my implementation and may need to throw all of this away 😭 

An example of the built-in a11y is that a `button`'s `onClick` is automatically triggered by the user hitting "enter" or "space" (and other inputs for screen readers). However, the `Menu` is forcibly stopping propagation, disrupting how the base `button` works. So, I'm going to play with it a bit more, but I think I should throw this new component.

**Comment by Justin-Avaya** [comment]
** Possibly out of scope
Any reason for three different hover treatments?
- Hover over Select all section is a slow fade in (too slow in my opinion)
- Hover over rows seems instant
- Hover over column header names is a sweeping animation

Example:
https://github.com/avaya-dux/neo-react-library/assets/161497909/45ce25a9-5010-42d1-927e-3910e9a60bd3

<hr>

** Possibly out of scope
As the rows are selected with the Select All checkbox, the columns shift/jump around. Selecting individual rows does not seem to have this effect with the exception of selecting the 4th row (Madam Hailey) which also creates the same shifting/jumping. This could be a row issue and not the Select All feature

Example:
https://github.com/avaya-dux/neo-react-library/assets/161497909/8540692f-ec2e-44eb-8f65-e725ccfc02ba

<hr>

Vertical sizing seems off from the source. The Figma document indicates the vertical height should be 44px whereas the story shows a vertical height of 61px

Example:
Screenshot 2024-07-09 at 12 41 05 PM

Screenshot 2024-07-09 at 12 43 10 PM

<hr>

If all of the above are out of scope and need to be addressed elsewhere, just let me know. The rest of it looked fine.

**Comment by joe-s-avaya** [comment]
Need to do some refactoring, closing for now and will re-open.

**Comment by joe-s-avaya** [comment]
> **Possibly out of scope**

@Justin-Avaya, this is all out of scope for this PR as my changes did not cause this. However, this many inconsistencies is particularly worrying and I've thus made a [jira issue](https://jira.forge.avaya.com/browse/NEO-2316) to resolve these. I'll bring this up with the team in standup tomorrow so we can schedule it.

**Comment by joe-s-avaya** [comment]
> functional
> Can not move focus to the last menu item with keyboard.

@yangAtSpoken can't reproduce. I must have broken it and then fixed it all without realizing it 😅

**Review by Justin-Avaya — APPROVED** [review]
- [x] "Select all" checkbox appropriately appears and updates.
- [x] Chevron next to the checkbox is visually appropriate (sizing, spacing, etc.).

✅ Approved

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 2 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: Excellent call! As they say, naming is the hardest problem in software, and your recommendations are a significant improvement; thanks for being thorough! I have made the updates that you recommended.

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 3 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: updated

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 4 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: updated

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 5 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: updated

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 6 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: updated

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 7 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: updated

File: src/components/Table/TableComponents/TableHeader.tsx (hunk 8 above)
Comment (joe-s-avaya) [reply src/components/Table/TableComponents/TableHeader.tsx]: updated

**Comment by enrique-prado** [comment]
From a UX perspective I think we are missing a "deselect all rows" option. If the user selects all rows and then decides to unselect them then he/she is forced to go to every page and deselect them. For tables with many pages that's not very usable.

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 10]
```diff
@@ -69,19 +87,71 @@ export const TableHeader = <T extends Record<string, any>>({
...
+											toggleAllRows();
+										}}
+									>
+										{allRowsAreSelected ? (
+											<>
+												{translations.clearSelection} ({rows.length})
+											</>
+										) : (
+											<>
+												{translations.selectAll} ({rows.length})
```
Comment (enrique-prado) [inline src/components/Table/TableComponents/TableHeader.tsx:149]: We also need an menu option to clear all selected rows. See my comments about UX in this PR.

File: src/components/Table/helpers/default-data.ts [hunk 11]
```diff
@@ -20,7 +20,10 @@ export const translations: ITableTranslations = {
 	header: {
 		clearSort: "Clear Sort",
 		filterColumn: "Filter Column",
-		selectAll: "select all",
+		clearSelection: "Clear all selections",
```
Comment (enrique-prado) [inline src/components/Table/helpers/default-data.ts:23]: I never saw this string in all my testing. Is it missing from your logic? Perhaps a bug?

**Comment by matthewAvaya** [comment]
- After selecting **Select all items (10)** from the header dropdown, the banner to **clear all selections** is missing. This light blue banner is missing in some of the other instances. 
Screen Shot 2024-07-10 at 2 34 03 PM

**Review by enrique-prado — COMMENTED** [review]
There seems to be an issue with the "Clear all selections" option. All my comments are related to that.

**Comment by joe-s-avaya** [comment]
@matthewAvaya 

> * After selecting **Select all items (10)** from the header dropdown, the banner to **clear all selections** is missing. This light blue banner is displayed anytime a user selects one of the dropdown items from the header dropdown of the checkbox.

This is not included in this PR. It will be added in the next one. 

> * Change copy **Select all items** to **Select all items in the table**

Updated

> * Change copy **Clear all selections** to **Clear all items**

Updated

**Review by yangAtSpoken — APPROVED** [review]
Looks good to me.

**Comment by matthewAvaya** [comment]
> @matthewAvaya
> 
> > * After selecting **Select all items (10)** from the header dropdown, the banner to **clear all selections** is missing. This light blue banner is displayed anytime a user selects one of the dropdown items from the header dropdown of the checkbox.
> 
> This is not included in this PR. It will be added in the next one.
> 
> > * Change copy **Select all items** to **Select all items in the table**
> 
> Updated
> 
> > * Change copy **Clear all selections** to **Clear all items**
> 
> Updated

I like the new copy that you added to the dropdown. 
Screen Shot 2024-07-11 at 10 32 55 AM

For clearing, let's use a similar copy. 

- Clear all items on page (5)
- Clear all items in the table (10)

**Review by matthewAvaya — APPROVED** [review]

**Pull request merged**